presto-main/src/main/java/com/facebook/presto/operator/MergeHashSort.java

```diff
@@ -0,0 +1,197 @@
+        for (int i = 0; i < keyTypes.size(); i++) {
+            hashChannels[i] = i;
+        }
+        HashGenerator hashGenerator = new InterpretedHashGenerator(keyTypes, hashChannels);
+
+        return new PageRewriteIterator(
+                hashGenerator,
+                allTypes,
+                Iterators.mergeSorted(
+                        channelIterators,
+                        (PagePosition left, PagePosition right) -> comparePages(hashGenerator, left, right)));
+    }
+
+    private static int comparePages(HashGenerator hashGenerator, PagePosition left, PagePosition right)
+    {
+        if (left.isPositionOutOfPage() && right.isPositionOutOfPage()) {
+            return 0;
+        }
+        if (left.isPositionOutOfPage()) {
+            return -1;
+        }
+        if (right.isPositionOutOfPage()) {
+            return 1;
+        }
+
+        long leftHash = hashGenerator.hashPosition(left.getPosition(), left.getPage());
+        long rightHash = hashGenerator.hashPosition(right.getPosition(), right.getPage());
+
+        return Long.compare(leftHash, rightHash);
+    }
+
+    public static class PagePosition
+    {
+        private final Page page;
+        private final int position;
+
+        public PagePosition(Page page, int position)
+        {
+            this.page = requireNonNull(page, "page is null");
+            this.position = requireNonNull(position, "position is null");
+        }
+
+        public Page getPage()
+        {
+            return page;
+        }
+
+        public int getPosition()
+        {
+            return position;
+        }
+
+        public boolean isPositionOutOfPage()
+        {
+            return position >= page.getPositionCount();
+        }
+    }
+
+    public static interface PagePositions extends Iterator<PagePosition>
+    {
+    }
+
+    public static class SingleChannelPagePositions
+            implements PagePositions
+    {
+        private final Iterator<Page> channel;
+        private PagePosition current;
+
+        public SingleChannelPagePositions(Iterator<Page> channel)
+        {
+            this.channel = requireNonNull(channel, "channel is null");
+        }
+
+        @Override
+        public boolean hasNext()
+        {
+            return channel.hasNext() || (current != null && current.getPosition() + 1 < current.getPage().getPositionCount());
+        }
+
+        @Override
+        public PagePosition next()
+        {
+            if (current == null || current.getPosition() + 1 >= current.getPage().getPositionCount()) {
+                current = new PagePosition(channel.next(), 0);
+            }
+            else {
+                current = new PagePosition(current.getPage(), current.getPosition() + 1);
+            }
+            return current;
+        }
+    }
+
+    /**
+     * This class rewrites iterator over PagePosition to iterator over Pages.
+     * Positions with same hash value are guaranteed to be in the same result page.
```

Comment:
I disagree, I think this could easily occur in real-world use. For example, suppose you're performing a MAX aggregation on varchars, or any other aggregation that can have large state objects. If you don't have enough memory and have to spill, there could easily end up being a very large number of pages that then have to be merged together.

If it's easiest to keep things in single pages though, we can do that in theory, but it means that this PageBuilder and its Pages need memory tracking to avoid running the VM out of memory. That seems like a more complicated change to me, and also error prone as we generally don't track single Page objects, as they're assumed to be small